src/TeenyStatistics.ts

```diff
@@ -0,0 +1,191 @@
+/*!
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export interface TeenyStatisticsOptions {
+  /**
+   * A positive number representing when to issue a warning about the number
+   * of concurrent requests using teeny-request.
+   * Set to 0 to disable this warning.
+   * Corresponds to the TEENY_REQUEST_WARN_CONCURRENT_REQUESTS environment
+   * variable.
+   */
+  concurrentRequests?: number;
+}
+
+type TeenyStatisticsConfig = Required<TeenyStatisticsOptions>;
+
+/**
+ * TeenyStatisticsCounters is distinct from TeenyStatisticsOptions:
+ * Used when dumping current counters and other internal metrics.
+ */
+export interface TeenyStatisticsCounters {
+  concurrentRequests: number;
+}
+
+/**
+ * @class TeenyStatisticsWarning
+ * @extends Error
+ * @description While an error, is used for emitting warnings when
+ *   meeting certain configured thresholds.
+ * @see process.emitWarning
+ */
+export class TeenyStatisticsWarning extends Error {
+  static readonly CONCURRENT_REQUESTS = 'ConcurrentRequestsExceededWarning';
+
+  public threshold = 0;
+  public type = '';
+  public value = 0;
+
+  /**
+   * @param {string} message
+   */
+  constructor(message: string) {
+    super(message);
+    this.name = this.constructor.name;
+    Error.captureStackTrace(this, this.constructor);
+  }
+}
+
+/**
+ * @class TeenyStatistics
+ * @description Maintain various statistics internal to teeny-request. Tracking
+ *   is not automatic and must be instrumented within teeny-request.
+ */
+export class TeenyStatistics {
+  /**
+   * @description A default threshold representing when to warn about excessive
+   *   in-flight/concurrent requests.
+   * @type {number}
+   * @static
+   * @readonly
+   * @default 5000
+   */
+  static readonly DEFAULT_WARN_CONCURRENT_REQUESTS = 5e3;
```

Comment:
I'd probably just make this `5000`, vs., using scientific notation (just a little more readable).

Reply:
fixed, using `5000` now.